File: articles/personality-and-physical-health-across-the-lifespan/python/analyze_personality_physical_health.py
#!/usr/bin/env python3
"""Analyze synthetic personality and physical health data.

This script supports the article "Personality and Physical Health Across the
Lifespan." It demonstrates age-band summaries, life-context summaries,
trait-health correlations, behavior / stress indices, and simple regression-style
models using synthetic data only.
"""

from __future__ import annotations

from pathlib import Path
import json
import numpy as np
import pandas as pd


ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "synthetic_personality_physical_health_lifespan.csv"
OUTPUTS = ROOT / "outputs"

TRAITS = [
    "extraversion",
    "agreeableness",
    "conscientiousness",
    "neuroticism",
    "openness",
    "emotional_stability",
]

BEHAVIORS = [
    "exercise",
    "sleep_quality",
    "smoking_risk",
    "alcohol_risk",
    "medication_adherence",
]

OUTCOMES = [
    "physical_health_score",
    "functional_ability",
    "chronic_condition_burden",
]


def fit_ols(df: pd.DataFrame, y_col: str, x_cols: list[str]) -> dict[str, float]:
    """Fit a small OLS model with numpy and return coefficients."""
    X = df[x_cols].to_numpy(dtype=float)
    X = np.column_stack([np.ones(len(X)), X])
    y = df[y_col].to_numpy(dtype=float)
    beta, *_ = np.linalg.lstsq(X, y, rcond=None)
    names = ["intercept", *x_cols]
    return {name: float(value) for name, value in zip(names, beta)}


def main() -> None:
    OUTPUTS.mkdir(parents=True, exist_ok=True)

    df = pd.read_csv(DATA_PATH)

    required = {
        "person_id",
        "wave",
        "age",
        "age_band",
        "life_context",
        *TRAITS,
        "perceived_support",
        *BEHAVIORS,
        "stress_burden",
        *OUTCOMES,
    }
    missing = required.difference(df.columns)
    if missing:
        raise ValueError(f"Missing required columns: {sorted(missing)}")

    df["health_behavior_index"] = (
        df["exercise"]
        + df["sleep_quality"]
        + df["medication_adherence"]
        - df["smoking_risk"]
        - df["alcohol_risk"]
    ) / 3.0

    df["stress_vulnerability_index"] = (
        df["neuroticism"]
        + df["stress_burden"]
        - df["emotional_stability"]
        - df["perceived_support"]
    )

    df["healthy_aging_support_index"] = (
        df["physical_health_score"]
        + df["functional_ability"]
        + df["perceived_support"]
        + df["medication_adherence"]
    ) / 4.0

    df["healthy_neuroticism_index"] = (
        df["neuroticism"] * df["conscientiousness"]
    ) / 7.0

    age_summary = (
        df.groupby("age_band")
        .agg(
            observations=("person_id", "count"),
            mean_age=("age", "mean"),
            physical_health_mean=("physical_health_score", "mean"),
            functional_ability_mean=("functional_ability", "mean"),
            chronic_condition_burden_mean=("chronic_condition_burden", "mean"),
            health_behavior_mean=("health_behavior_index", "mean"),
            stress_vulnerability_mean=("stress_vulnerability_index", "mean"),
        )
        .reset_index()
        .sort_values("mean_age")
    )

    context_summary = (
        df.groupby("life_context")
        .agg(
            observations=("person_id", "count"),
            perceived_support_mean=("perceived_support", "mean"),
            medication_adherence_mean=("medication_adherence", "mean"),
            stress_burden_mean=("stress_burden", "mean"),
            physical_health_mean=("physical_health_score", "mean"),
            functional_ability_mean=("functional_ability", "mean"),
        )
        .reset_index()
    )

    age_summary.to_csv(OUTPUTS / "python_age_band_summary.csv", index=False)
    context_summary.to_csv(OUTPUTS / "python_life_context_summary.csv", index=False)

    corr_cols = TRAITS + [
        "perceived_support",
        "stress_burden",
        "health_behavior_index",
        "stress_vulnerability_index",
        "healthy_aging_support_index",
        "healthy_neuroticism_index",
    ] + BEHAVIORS + OUTCOMES

    df[corr_cols].corr().to_csv(OUTPUTS / "python_trait_health_correlations.csv")

    health_inputs = [
        "conscientiousness",
        "neuroticism",
        "extraversion",
        "openness",
        "exercise",
        "sleep_quality",
        "smoking_risk",
        "medication_adherence",
        "stress_burden",
        "age",
        "wave",
    ]

    healthy_neuroticism_inputs = [
        "neuroticism",
        "conscientiousness",
        "healthy_neuroticism_index",
        "exercise",
        "sleep_quality",
        "smoking_risk",
        "medication_adherence",
        "stress_burden",
        "age",
        "wave",
    ]

    function_inputs = [
        "physical_health_score",
        "conscientiousness",
        "emotional_stability",
        "perceived_support",
        "medication_adherence",
        "age",
        "wave",
    ]

    chronic_inputs = [
        "age",
        "stress_burden",
        "smoking_risk",
        "alcohol_risk",
        "conscientiousness",
        "physical_health_score",
        "perceived_support",
    ]

    models = {
        "physical_health_score": fit_ols(df, "physical_health_score", health_inputs),
        "healthy_neuroticism_model": fit_ols(df, "physical_health_score", healthy_neuroticism_inputs),
        "functional_ability": fit_ols(df, "functional_ability", function_inputs),
        "chronic_condition_burden": fit_ols(df, "chronic_condition_burden", chronic_inputs),
        "note": "Synthetic demonstration only; not a clinical prediction or public-health scoring model.",
    }

    with open(OUTPUTS / "python_personality_health_models.json", "w", encoding="utf-8") as f:
        json.dump(models, f, indent=2)

    df.to_csv(OUTPUTS / "python_scored_personality_physical_health.csv", index=False)

    print("Wrote Python outputs to:", OUTPUTS)


if __name__ == "__main__":
    main()

File: articles/personality-and-physical-health-across-the-lifespan/r/analyze_personality_physical_health.R
# Personality and Physical Health Across the Lifespan
# R workflow for longitudinal personality-health analysis

root <- normalizePath(file.path(dirname(sys.frame(1)$ofile), ".."))
data_path <- file.path(root, "data", "synthetic_personality_physical_health_lifespan.csv")
outputs <- file.path(root, "outputs")
dir.create(outputs, showWarnings = FALSE, recursive = TRUE)

data <- read.csv(data_path, stringsAsFactors = FALSE)

required <- c(
  "person_id",
  "wave",
  "age",
  "age_band",
  "life_context",
  "extraversion",
  "agreeableness",
  "conscientiousness",
  "neuroticism",
  "openness",
  "emotional_stability",
  "perceived_support",
  "exercise",
  "sleep_quality",
  "smoking_risk",
  "alcohol_risk",
  "medication_adherence",
  "stress_burden",
  "physical_health_score",
  "functional_ability",
  "chronic_condition_burden"
)

missing <- setdiff(required, names(data))
if (length(missing) > 0) {
  stop(paste("Missing required columns:", paste(missing, collapse = ", ")))
}

data$health_behavior_index <- (
  data$exercise +
  data$sleep_quality +
  data$medication_adherence -
  data$smoking_risk -
  data$alcohol_risk
) / 3

data$stress_vulnerability_index <- (
  data$neuroticism +
  data$stress_burden -
  data$emotional_stability -
  data$perceived_support
)

data$healthy_aging_support_index <- (
  data$physical_health_score +
  data$functional_ability +
  data$perceived_support +
  data$medication_adherence
) / 4

data$healthy_neuroticism_index <- (
  data$neuroticism * data$conscientiousness
) / 7

age_summary <- aggregate(
  data[, c("physical_health_score", "functional_ability", "chronic_condition_burden", "health_behavior_index", "stress_vulnerability_index")],
  by = list(age_band = data$age_band),
  FUN = mean
)

age_n <- aggregate(
  data$person_id,
  by = list(age_band = data$age_band),
  FUN = length
)
names(age_n)[2] <- "observations"
age_summary <- merge(age_n, age_summary, by = "age_band")

write.csv(age_summary, file.path(outputs, "r_age_band_summary.csv"), row.names = FALSE)

corr_cols <- c(
  "extraversion",
  "agreeableness",
  "conscientiousness",
  "neuroticism",
  "openness",
  "emotional_stability",
  "perceived_support",
  "exercise",
  "sleep_quality",
  "smoking_risk",
  "alcohol_risk",
  "medication_adherence",
  "stress_burden",
  "physical_health_score",
  "functional_ability",
  "chronic_condition_burden"
)

cor_matrix <- cor(data[, corr_cols], use = "pairwise.complete.obs")
write.csv(cor_matrix, file.path(outputs, "r_trait_health_correlations.csv"))

model_health <- lm(
  physical_health_score ~ conscientiousness + neuroticism + extraversion +
    openness + exercise + sleep_quality + smoking_risk +
    medication_adherence + stress_burden + age + wave,
  data = data
)

model_healthy_neuroticism <- lm(
  physical_health_score ~ neuroticism * conscientiousness +
    exercise + sleep_quality + smoking_risk +
    medication_adherence + stress_burden + age + wave,
  data = data
)

model_function <- lm(
  functional_ability ~ physical_health_score + conscientiousness +
    emotional_stability + perceived_support + medication_adherence +
    age + wave,
  data = data
)

model_chronic <- lm(
  chronic_condition_burden ~ age + stress_burden + smoking_risk +
    alcohol_risk + conscientiousness + physical_health_score +
    perceived_support,
  data = data
)

capture.output(summary(model_health), file = file.path(outputs, "r_physical_health_model_summary.txt"))
capture.output(summary(model_healthy_neuroticism), file = file.path(outputs, "r_healthy_neuroticism_model_summary.txt"))
capture.output(summary(model_function), file = file.path(outputs, "r_functional_ability_model_summary.txt"))
capture.output(summary(model_chronic), file = file.path(outputs, "r_chronic_condition_model_summary.txt"))

write.csv(data, file.path(outputs, "r_scored_personality_physical_health.csv"), row.names = FALSE)

cat("Wrote R outputs to:", outputs, "\n")
